posts/alpine-stack-size.md

```diff
@@ -0,0 +1,172 @@
+
+```
+SystemStackError: stack level too deep
+```
+
+**What we knew now:**
+
+* The stack size was being restricted within the Alpine docker image
+* The `RUBY_THREAD_VM_STACK_SIZE` environment variable had no affect within this docker image
+
+## What was wrong?
+
+Doing some google searching we ended up at musl-libc's FAQs, specifically the ["functional differences"](https://wiki.musl-libc.org/functional-differences-from-glibc.html#Thread-stack-size) i.e. incompatibilities with glibc. 
+
+What we learned is that musl provides a default stack size of 80k, which is a significant difference to the stacksize provided by glibc (generally 2MB-10MB).
+
+Moreover, it requires programs to explicitly ask for more (or less) stack via `pthread_attr_setstacksize`, which MRI Ruby does not appear to do.
+
+There is an open Ruby bug here: https://bugs.ruby-lang.org/issues/14387
+
+### What is RSpec doing?
+
+RSpec prints out a list of errors, and helpfully includes a snippet of source code from your test suite.
+
+```
+Failures:
+
+  1) 1 equals 2
+     Failure/Error: expect(1).to eq(2)
+
+       expected: 2
+            got: 1
+
+       (compared using ==)
+     # ./spec/test_spec.rb:12:in `block (11 levels) in <top (required)>'
+
+Finished in 0.01589 seconds (files took 0.07908 seconds to load)
+1 example, 1 failure
+
+Failed examples:
+
+rspec ./spec/test_spec.rb:11 # 1 equals 2
+```
+
+In the example above, the string `expect(1).to eq(2)` is extracted from your test suite source code.
+
+The way it does is that it parses your Ruby source code using [Ruby ripper](http://ruby-doc.org/stdlib-2.5.0/libdoc/ripper/rdoc/Ripper.html), and then searches for the relevant nodes in the Ruby language Abstract Syntax Tree (AST). 
+
+This requires a Depth-First-Search algorithm, which is more elegant when expressed as a recursive algorithm.
+
+## How did we solve it?
+
+We submitted a [patch upstream to RSpec](https://github.com/rspec/rspec-support/pull/343) to change the implementation from a recursive algorithm to a loop-based algorithm that does not use the stack.
+
+Ultimately this is a temporary fix, as any recursive logic could be a source of errors in production. This should be solved at the Ruby language level.
+
+## The whiff of a larger design smell
+
+This is a Liskov-substitution principle violation.
+
+*Liskov substituion principle?? That's an object-oriented principle? How does a stack overflow error caused by libc relate to Objects?*
+
+The dependency of what libc implementation is used is controlled by the Linux distribution (the caller of your Ruby application). This means the dependency is inverted, and we are looking at an object-oriented design.
+
+If you consider `musl-libc` and `glibc` as implementing an interface called `libc`, then you will notice that they must also be Liskov-compatible to ensure smooth operation.
+
+The imperfection here is that their public APIs are slightly different, possibly by mistake, or probably by choice.
```

Comment:
Wonder whether to include a bit of history about Alpine's background with LEAF (which is a distribution designed for embedded systems, which needs the small stacksize).